src/zql/ast-to-ivm/README.md

```diff
@@ -0,0 +1,45 @@
+# ast-to-ivm
+
+Given an AST (as defined by `ZqlAst.ts`), build up an IVM pipeline.
+
+There are a few stages towards implementation here --
+
+# Implementation Stages
+
+## Stage 1: Pipeline Per ZQL Query
+
+This is the simplest conversion. Each ZQL query gets a unique pipeline. This means if we have duplicate queries then we get duplicate pipelines.
+
+E.g.,
+
+```ts
+for (let i = 0; i < 1000; ++i) {
+  Issue.select('id').prepare();
+}
+```
+
+Would prepare 1k identical pipelines and a write would visit all 1k pipelines rather than the statement being collapsed to a single ref-counted pipeline.
+
+## Stage 2: Pipeline per Unique Bound ZQL Query
+
+This ref counts so identical queries (with identical bindings) share the same pipeline. Identical queries with differing bindings still get unique pipelines.
```

Comment:
I thought a bit about this and I think you do not need ref counting and would be better off using `WeakRef`s. But I don't have the full picture in my head at this point.

Writing down my assumptions so I do not forget this again.

- Identical queries compute the same hash (string or number).
- There is a mapping from the hash to the shared object.
- All the instances have a strong ref to the shared object
- The shared object is wrapped in a WeakRef so that when all the strong refs to the shared object are GC'ed we can remove the hash from the mapping.
- We compute a hash representation of a query. ie. #deadbeef.

Reply:
I'll keep this in mind once we get to de-duplicating queries.